### PR TITLE
adds volatile state directory to the fs plan for cntrs/pods/fifo

### PR DESCRIPTION
**Pull request by mikebrow**
Addressing issue: #686

Signed-off-by: Mike Brown <brownwm@us.ibm.com>

### CODE REVIEW

**Comment by mikebrow** [comment]
rebased

File: docs/config.md [hunk 1]
```diff
@@ -38,9 +38,12 @@ The explanation and default value of each configuration item are as follows:
       # runtime_engine is the name of the runtime engine used by containerd.
       runtime_engine = ""
 
-      # runtime_root is the directory used by containerd for runtime state.
+      # runtime_root is the directory used by containerd for persistent runtime state.
       runtime_root = ""
 
+      # runtime_state is the directory used by containerd for volatile runtime state.
+      runtime_state = ""
```
Comment (Random-Liu) [inline docs/config.md]: Why do we need this? We are not using it.

Comment (Random-Liu) [reply docs/config.md]: There is no such thing in runc options, either.
https://github.com/containerd/containerd/blob/856ea51c8b5923b88ae41c9df501909c07776ea7/linux/runctypes/runc.pb.go#L40

File: hack/test-integration.sh [hunk 2]
```diff
@@ -26,14 +26,16 @@ FOCUS=${FOCUS:-""}
...
 
 mkdir -p ${REPORT_DIR}
 test_setup ${REPORT_DIR}
 
 # Run integration test.
 sudo ${ROOT}/_output/integration.test --test.run="${FOCUS}" --test.v \
   --cri-endpoint=${CONTAINERD_SOCK} \
-  --cri-root=${CRI_ROOT}
+  --cri-root=${CRI_ROOT} \
+  --cri-state=${CRI_STATE}
```
Comment (Random-Liu) [inline hack/test-integration.sh]: We don't need this for now, volume should be in persistent state.

File: integration/volume_copy_up_test.go [hunk 3]
```diff
@@ -70,7 +70,7 @@ func TestVolumeCopyUp(t *testing.T) {
 	assert.Equal(t, "test_content\n", string(stdout))
 
 	t.Logf("Check host path of the volume")
-	hostCmd := fmt.Sprintf("ls %s/containers/%s/volumes/*/test_file | xargs cat", *criRoot, cn)
+	hostCmd := fmt.Sprintf("ls %s/containers/%s/volumes/*/test_file | xargs cat", *criState, cn)
```
Comment (Random-Liu) [inline integration/volume_copy_up_test.go]: Volume should be in persistent state.

How about we add "volatileStateRoot" and only change logic for `shm` and `fifos`, and keep other original logic.

File: pkg/server/container_create.go [hunk 4]
```diff
@@ -134,26 +134,35 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
...
+	}
 	defer func() {
 		if retErr != nil {
-			// Cleanup the container root directory.
+			// Cleanup the container root directories.
 			if err = c.os.RemoveAll(containerRootDir); err != nil {
 				logrus.WithError(err).Errorf("Failed to remove container root directory %q",
 					containerRootDir)
 			}
+			if err = c.os.RemoveAll(persistentContainerRootDir); err != nil {
```
Comment (Random-Liu) [inline pkg/server/container_create.go]: Each defer cleanup should be right after the creation.
e.g. `containerRootDir` cleanup won't run when `c.os.MkdirAll(persistentContainerRootDir, 0755)` returns error, which is not what we want.

File: pkg/server/sandbox_run.go [hunk 5]
```diff
@@ -189,19 +189,28 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
...
 	}
+	persistentSandboxRootDir := getSandboxRootDir(c.config.RootDir, id)
+	if err := c.os.MkdirAll(persistentSandboxRootDir, 0755); err != nil {
+		return nil, errors.Wrapf(err, "failed to create persistent sandbox root directory %q",
+			persistentSandboxRootDir)
+	}
 	defer func() {
 		if retErr != nil {
-			// Cleanup the sandbox root directory.
+			// Cleanup the sandbox root directories.
```
Comment (Random-Liu) [inline pkg/server/sandbox_run.go]: Same defer issue.

File: pkg/server/sandbox_run.go [hunk 6]
```diff
@@ -189,19 +189,28 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		}
 	}()
 
-	// Create sandbox container root directory.
-	sandboxRootDir := getSandboxRootDir(c.config.RootDir, id)
+	// Create sandbox container root directories.
+	sandboxRootDir := getSandboxRootDir(c.config.StateDir, id)
```
Comment (Random-Liu) [inline pkg/server/sandbox_run.go]: `hosts` and `resolv.conf` should also be persistent.

**Comment by Random-Liu** [comment]
I think at least till now, we have more things in "persistent" directly. Let's still call it `rootDir`, and add a `volatileStateRootDir`. We only need fifos and `shm` in `volatileStateRootDir`.

**Comment by Random-Liu** [comment]
@mikebrow It seems that you are not online today. We need to get PR merged today. :)

I carried on this PR in https://github.com/containerd/cri/pull/697.

File: docs/config.md (hunk 1 above)
Comment (mikebrow) [reply docs/config.md]: yeah it's a todo..

**Pull request closed without merging**